testing.md

```diff
@@ -0,0 +1,299 @@
+---
+layout: page
+title: Testing
+permalink: /testing/
+---
+
+* TOC
+{:toc}
+
+## Introduction
+
+The more features your application has, the harder it is to test it manually. A
+good set of automated tests will help you make sure your application performs
+correctly when you publish it to users while retaining your feature and bug fix
+velocity.
+
+There are many kinds of automated testing. Their definitions are fuzzy, so let's
+define them here for clarity:
+
+- A _unit test_ tests a single function, method or class. External dependencies
+  of the test subject are generally mocked out using, for example,
+  [`package:mockito`](https://github.com/fibulwinter/dart-mockito/blob/master/README.md).
+  Unit tests generally do not read from/write to disk, render to screen and do
+  not receive user actions from outside the process running the test. The goal
+  of a unit-test is to verify the correctness of a piece of logic under a
+  variety of conditions.
+- A _widget test_ (in other UI frameworks referred to as _component test_) tests
+  a single widget. Testing a widget involves multiple classes and requires a
+  test environment that makes the widget feel like it's part of a real UI. For
+  example, it should be able to receive and respond to user actions and events,
+  perform layout, and instantiate child widgets. A widget test is
+  therefore more comprehensive than a unit test. However, like a unit test, a
+  widget test's environment is replaced with an implementation much simpler than
+  a full-blown UI system. The goal of a widget test is to verify that it looks
+  and interacts as expected.
+- An _end-to-end test_ tests a complete application or a large part of an
+  application. Generally an _end-to-end test_ runs on a real device or an OS
+  emulator. The test code does not run in the same process (or even the same
+  device) as the test subject. The goal of an end-to-end test is to verify that
+  the application functions correctly as a whole, that all the widgets it is
+  composed of integrate with each other as expected (which is why another term
+  for these kinds of tests is _integration test_). Another important aspect
+  end-to-end tests verify is application performance.
+
+Here is a table summarizing the tradeoffs concerning the choice between
+different kinds of tests:
+
+|                      | Unit   | Widget | End-to-end |
+|----------------------|--------|--------|------------|
+| **Effectiveness**    | Low    | Higher | Highest    |
+| **Maintenance cost** | Low    | Higher | Highest    |
+| **Dependencies**     | Few    | More   | Lots       |
+| **Execution speed**  | Quick  | Slower | Slowest    |
+
+As a rule of thumb a well-tested application has a very high number of unit
```

Comment:
This is a very good recommendation. Suggest highlighting it by adding a bold **Tip:** prefix (before "As a rule...").